fix: don't use logical diagnostic position in `get_prev` and `get_next`

faergeek opened the pull request:
Fixes #35136

Problem: `vim.diagnostic.get_prev()` / `vim.diagnostic.get_next()` use logical diagnostic positions and consider extmark validity. `vim.diagnostic.get()` only uses original positions and doesn't care about extmark validity. This results in inconsistency between these APIs.

Solution: use original positions in `vim.diagnostic.get_prev()` and `vim.diagnostic.get_next()` and don't consider extmark validity to match previous behavior, which is consistent with `vim.diagnostic.get`.

<!--
  Thank you for contributing to Neovim!
  If this is your first time, check out https://github.com/neovim/neovim/blob/master/CONTRIBUTING.md#pull-requests-prs
  for our PR guidelines.
-->

MariaSolOs commented on runtime/lua/vim/diagnostic.lua:
```diff
@@ -662,15 +662,24 @@ local function get_logical_pos(diagnostic)
+    local valid ---@type boolean
+
+    if use_logical_pos then
+      lnum, _, _, _, valid = get_logical_pos(diagnostic)
```
Unrelated but could we add a comment in `get_logical_pos` explaining what we mean by a diagnostic's logical position?

faergeek replied:
Added a comment

faergeek committed:
fix: don't use logical diagnostic position in `get_prev` and `get_next`

Fixes #35136

Problem: `vim.diagnostic.get_prev()` / `vim.diagnostic.get_next()` use
logical diagnostic positions and consider extmark validity.
`vim.diagnostic.get()` only uses original positions and doesn't care
about extmark validity. This results in inconsistency between these
APIs.

Solution: use original positions in `vim.diagnostic.get_prev()` and
`vim.diagnostic.get_next()` and don't consider extmark validity to match
previous behavior, which is consistent with `vim.diagnostic.get`.
